api/authentication.py

```diff
@@ -10,77 +10,101 @@
 
 from allauth.socialaccount.models import SocialAccount
 from rest_framework.authentication import BaseAuthentication, get_authorization_header
+from rest_framework.exceptions import AuthenticationFailed
 
 
 logger = logging.getLogger("events")
 
 
 def get_cache_key(token):
+    # TODO: token can be a JWT which makes for a really big cache key
     return f"fxa_token_{token}"
 
 
+def introspect_token(introspect_token_url, token, cache_key, cache_timeout):
+    try:
+        fxa_resp = requests.post(introspect_token_url, json={"token": token})
+    except:
+        logger.error(
+            "Could not introspect token with FXA.",
+            extra={"fxa_response": shlex.quote(fxa_resp.text)},
+        )
+        # cache empty response data to prevent FXA failures from causing runaway
+        # HTTP retries
+        fxa_resp_data = {"status_code": None, "json": {}}
+        cache.set(cache_key, fxa_resp_data, cache_timeout)
```

Comment:
Nit: This code is repeated three times, since it is trying to ensure the cache is always, which is really a concern of the calling function. Here's what it might look like in ``FxaTokenAuthenication.authenticate``:

```python
            fxa_resp_data = None
            try:
                fxa_resp_data = introspect_token(introspect_token_url, token)
            except AuthenticationFailed:
                # cache empty response data to prevent FXA failures from causing runaway
                # HTTP retries
                fxa_resp_data = {"status_code": None, "json": {}}
                raise
            finally:
                cache.set(cache_key, fxa_resp_data, cache_timeout)

Reply:
Changed the calling function to include `finally: cache.set` thanks!